### ISSUE TITLE
Use *args, **kwargs with jit(static_argnums=...) for numpy.gradient

### ISSUE THREAD

**Issue by clemisch** [google/jax#1925]
Dear jax team,

I am implementing `numpy.gradient` in jax and, while my code numerically works, I am struggling with handling `static_argnums` where the static argument is a keyword argument.

The original function signature is `numpy.gradient(f, *varargs, **kwargs)` where `*varargs` are spacings and `**kwargs` may be `edge_order` or `axis`. I implemented only support for `axis`, but my question is valid for a complete implementation, too.

1. How can I follow numpy's signature while telling jax the static argument? 
2. Is following the signature strictly necessary when contributing to jax?

My implementation which works for me but does not follow the original signature:

```python
import jax
import jax.numpy as np
import numpy as onp


@jax.partial(jax.jit, static_argnums=1)
def gradient(a, axis=None):

  def _gradient_along_axis(a, axis):
    a_swap = np.swapaxes(a, 0, axis)
    a_grad = np.concatenate((
      (a_swap[1]  - a_swap[0])[np.newaxis],
      (a_swap[2:] - a_swap[:-2]) * 0.5,
      (a_swap[-1] - a_swap[-2])[np.newaxis]
      ), axis=0)
    return np.swapaxes(a_grad, 0, axis)

  if axis is None:
    axis = range(a.ndim)
  else:
    if isinstance(axis, int):
      axis = (axis,)
    if not isinstance(axis, tuple):
      raise ValueError("Give `axis` either as int or tuple of ints")

  a_grad = [_gradient_along_axis(a, ax) for ax in axis]

  if len(axis) == 1:
    a_grad = a_grad[0]

  return a_grad


# TEST
key = jax.random.PRNGKey(0)
x = jax.random.normal(key, (10, 20, 30))

for axis in [None, 0, 1, 2, -0, -1, -2, (0, 1), (0, 1, 2), (-1, -2), (-2, -1), (-1, 0)]:
  print(axis)
  assert np.allclose(
    gradient(x, axis),
    onp.gradient(x, axis=axis),
    rtol=1e-6, atol=1e-6
  )
```

**Comment by shoyer** [comment]
The standard way to handle keyword arguments or unusual function signatures with `jit` is to write a helper function, e.g.,
```python
@functools.partial(jax.jit, static_argnums=(1,))
def _my_func(x, y):
  ...

def my_func(x, y):
  return _my_func(x, y)
```

Now you can write something like `my_func(x=1, y=2)` and `y` will still be a static argument.

**Comment by clemisch** [comment]
Thank you! 

`my_func` won't be jit'able directly, right? If the output shape of `my_shape` fundamentally depends on `y`, that is.

So it is not a problem if `jax.numpy.gradient` would not be jit'able directly? (I was so focused on getting the function to jit that I overlooked that possibility.)

**Comment by shoyer** [comment]
> `my_func` won't be jit'able directly, right? If the output shape of `my_shape` fundamentally depends on `y`, that is.

Right, but you could `jit` a function that call `my_func` *inside* it, e.g.,
```
@jit
def other_func(x)
  return my_func(x, y=2)
```

> So it is not a problem if `jax.numpy.gradient` would not be jit'able directly? (I was so focused on getting the function to jit that I overlooked that possibility.)

Correct, this is not a problem.

**Comment by clemisch** [comment]
Great, thank you! I will implement tests with said functions and create a PR.
